docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+
+- in case of a fork, we accept that the handler temporarily stores
+  headers (tagged as verified). 
+  
+- eventually, a handler should be informed
+ (`checkMisbehaviourAndUpdateState`)
+ by some relayer that it has
+  verified a header from a fork. Then the handler should do what is
+ required by IBC in this case (stop?)
+
+### Challenges in the handler requirement
+
+- handlers and relayers work on different lightstores. In principle
+  the lightstore need not intersect in any heights a priori
+
+- if a relayer  sees a header *h* it doesn't know at a handler (`queryChainConsensusState`), the
+  relayer needs to
+  verify that header. If it cannot do it locally based on downloaded
+  and verified (trusted?) light blocks, it might need to use
+  `VerifyToTarget` (bisection). To call `VerifyToTarget` we might keep
+  *h* in the lightstore. If verification fails, we need to download the
+  "alternative" header of height *h.Height* to generate evidence for
+  the handler.
+  
+- we have to specify what precisely `queryChainConsensusState`
+  returns. It cannot be the complete lightstore. Is the last header enough?
+
+- we would like to assume that every now and then (smaller than the
+  trusting period) a correct relayer checks whether the handler is on a
+  different branch than the relayer. 
+  And we would like that this is enough to achieve
+  the Handler requirement.
+  
+   - here the correctness argument would be easy if a correct relayer is
+     based on a light client with a *trusted* state, that is, a light
+     client who never changes its opinion about trusted. Then if such a
+     correct relayer checks-in with a handler, it will detect a fork, and
+     act in time.
+
+   - if the light client does not provide this interface, in the case of
+     a fork, we need some assumption about a correct relayer being on a
+     different branch than the handler, and we need such a relayer to 
+	 check-in not too late. Also
+     what happens if the relayer's light client is forced to roll-back
+     its lightstore?
+     Does it have to re-check all handlers?
+
+
+
+
+## On the interconnectedness of things
+
+In the broader discussion of so-called "fork accountability" there are
+several subproblems
+
+- Fork detection 
+
+- Evidence creation and submission
+
+- Isolating misbehaving nodes (and report them for punishment over abci)
+
+
+### Fork detection
+
+The preliminary specification ./detection.md formalizes the notion of
+a fork. Roughly, a fork exists if there are two conflicting headers
+for the same height, where both are supported by bonded full nodes
+(that have been validators in the near past, that is, within the
+trusting period). We distinguish between *fork on the chain* where two
+conflicting blocks are signed by +2/3 of the validators of that
+height, and a *light client fork* where one of the conflicting headers
+is not signed by  +2/3 of the current height, but by +1/3 of the
+validators of some smaller height.
```

Comment:
This sentence is not clear. I am not sure distinction we make between on chain forks and light client forks is about voting power of faulty nodes. You can create a fork on the main chain with +1/3 of voting power. In my view it is more "location" of the fork which defines its type. Main chain fork is perceived by full nodes (including validators) by running Tendermint protocols and in this case we expect full node to panic and we enter off-chain fork accountability protocol. On the other side light client fork is detected by the light client and it enters Tendermint network sandboxed in an evidence, so we don't expect full node to panic in this case but to try to execute on chain fork accountability protocol triggered by this evidence. In case (due to censorship) we fail to execute evidence within some time frame, a full node (that is aware of evidence) should panic and resort to off chain fork accountability protocol. The intuition for this reasoning is that light client fork could be one-shot event (and therefore we try to process it) while the main chain fork is an attempt to take over main chain.

Reply:
For me, the distinction is not about the location. It is about "potential" location, if you want.

The Tendermint validation says that if +2/3 of the validators sign a block, then the block is OK, and all the protocols accept that.

If there are two blocks for the same height with +2/3 signatures, then both blocks are equally valid. Even if at the moment no full node is on an alternative branch, it might be that in the future, a node that does fast sync might accept it. Then it suddenly becomes a main chain fork. So a definition based on location does not seem stable. Therefore, I am not sure it is safe to make the distinction on the "location". This is also what underlines my axiomatization in the ./detection.md draft.

However, if there is a unique chain with +2/3 signatures then this chain is authoritative. If in addition there is a block that is signed by +1/3 of some older validator set, then the light client might accept it while skipping. However, by uniqueness, this cannot transform into a main chain fork in the future.

Reply:
Also, I believe if the bad guys have barely +1/3 in order to get +2/3 they have to act in the open, that is, on the systems, and it is likely that their attempt to fork the system can easily be observed/caught. This makes it unlikely to appear.

The light client attack with the +1/3 they can do among themselves, without exchanging messages in the open. This might be more likely to appear

Reply:
What I can't fully get is dynamics of +1/3 attack on the main chain where a different block is created for some height h. Is it possible to keep extending forked chain with +1/3 of voting power if in the worst case block at height h leads to changing val set such that correct validators are excluded and only faulty (+1/3) stays in?

Reply:
I guess yes. If in the other block the bad guys can control the next validator set, it might even be that the system falls apart. They should not have the same correct guy in both next validator sets. But otherwise, in principle, they could convince one correct validator to be validator in one branch and another correct validator to be validator in another branch. As agreement is violated, I guess this is possible.

Reply:
If I remember well, there is something about off by one that constrain scenarios. We elect new valset after block h is executed but they are in power only for block h+2. I was trying yesterday quickly to come up with a worst case scenario, but failed. But we can for now probably ignore this issue and try to figure out the rest :)

Reply:
> You can create a fork on the main chain with +1/3 of voting power. 

Yes but you still need +2/3 to sign the block.  For a light-client fork, that's not the case, you just need +1/3 of the old voting power to sign the block

Reply:
> What I can't fully get is dynamics of +1/3 attack on the main chain where a different block is created for some height h. Is it possible to keep extending forked chain with +1/3 of voting power if in the worst case block at height h leads to changing val set such that correct validators are excluded and only faulty (+1/3) stays in?

So to fork a full node, the two conflicting blocks at height H must have the same NextValidatorSet (validity condition). So if you make a fork on the main chain at height H with +1/3 faulty, both blocks at H+1 are going to have to have the same validator set, though they can now have different NextValidatorSet. It's only at H+2 that the Validator set itself can differ between the two forks. Does this mean that H+1 can't get committed and the fork stops after 1 block?

Reply:
I am a bit lost in indices, so to clarify:

- To create a fork on the chain, one needs an attack of two consecutive blocks.
- On height H we have our classic scenarios (equivocation, amnesia), 
- On height H+1 it basically means that the validators run two consensus instances for the same hight, which at the end also boils down to double signing, although they might behave "correctly" in each instance.

Is that right?

Reply:
> If I remember well, there is something about off by one that constrain scenarios. 
> Does this mean that H+1 can't get committed and the fork stops after 1 block?

Not sure about this, but it seems there's nothing preventing each branch of the fork to continue extending (to H+1 and beyond). For H+1, it will require the faulty guys to at least double sign, but by H+2, they can make the validator set change, and maintain the forks without more faulty behaviour.